src/ControlSystem/PiecewisePolynomial.hpp

```diff
@@ -0,0 +1,71 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <array>
+#include <vector>
+
+#include "DataStructures/DataVector.hpp"
+
+/// \ingroup ControlSystemGroup
+/// Contains functions of time to support the dual frame system.
+namespace FunctionsOfTime {
+/// \ingroup ControlSystemGroup
+/// A function that has a piecewise-constant `MaxDeriv`th derivative.
+template <size_t MaxDeriv>
+class PiecewisePolynomial {
+ public:
+  PiecewisePolynomial(
+      double t,
+      std::array<DataVector, MaxDeriv + 1> initial_func_and_derivs) noexcept;
+
+  ~PiecewisePolynomial() = default;
+  PiecewisePolynomial(PiecewisePolynomial&&) noexcept = default;
+  PiecewisePolynomial& operator=(PiecewisePolynomial&&) noexcept = default;
+  PiecewisePolynomial(const PiecewisePolynomial&) = delete;
+  PiecewisePolynomial& operator=(const PiecewisePolynomial&) = delete;
+
+  /// Returns the function and `MaxDerivReturned` derivatives at
+  /// an arbitrary time `t`.
+  /// The function has multiple components.
+  template <size_t MaxDerivReturned = MaxDeriv>
+  std::array<DataVector, MaxDerivReturned + 1> operator()(double t) const
+      noexcept;
+  /// Updates the `MaxDeriv`th derivative of the function at the given time.
+  /// updated_max_deriv is a vector of the `MaxDeriv`ths for each component
+  void update(double time_of_update, DataVector updated_max_deriv) noexcept;
+  /// Returns the domain of validity of the function.
+  std::array<double, 2> time_bounds() const noexcept {
+    return {{deriv_info_at_update_times_.front().time,
+             deriv_info_at_update_times_.back().time}};
+  }
+
+ private:
+  // There exists a DataVector for each deriv order that contains
+  // the values of that deriv order for all components.
+  using value_type = std::array<DataVector, MaxDeriv + 1>;
+
+  // Holds information at single time at which the `MaxDeriv`th
+  // derivative has been updated.
+  struct DerivInfo {
+    double time;
+    value_type derivs_coefs;
+    // Converts the derivs to coefficients for polynomial evaluation.
+    value_type convert_derivs_to_coefs(value_type deriv_coefs) const noexcept;
```

Comment:
here, eliminate this function; in the cpp file use its definition as the constructor definition